python/cudf/cudf/utils/_numba.py

```diff
@@ -0,0 +1,196 @@
+    with the user cuDF package.
+    """
+    try:
+        # By default, ptxcompiler will not be installed with CUDA 12
+        # packages. This is ok, because in this situation putting
+        # numba in enhanced compatibility mode is not necessary.
+        from ptxcompiler.patch import NO_DRIVER, safe_get_versions
+    except ModuleNotFoundError:
+        versions = _get_versions()
+        if versions != _NO_DRIVER:
+            driver_version, runtime_version = versions
+            if runtime_version > driver_version:
+                warnings.warn(
+                    f"Using CUDA toolkit version {runtime_version} with CUDA "
+                    f"driver version {driver_version} requires minor version "
+                    "compatibility, which is not yet supported for CUDA "
+                    "driver versions newer than 12.0. It is likely that many "
+                    "cuDF operations will not work in this state. Please "
+                    f"install CUDA toolkit version {driver_version} to "
+                    "continue using cuDF."
+                )
+        return
+    versions = safe_get_versions()
+    if versions != NO_DRIVER:
+        driver_version, runtime_version = versions
+        # Don't check if CEC is necessary in the possible edge
+        # case where a user has a CUDA 12 package and ptxcompiler
+        # in their environment anyways, perhaps installed separately
+        if driver_version < (12, 0):
+            ptx_toolkit_version = _get_cuda_version_from_ptx_file(
+                CC_60_PTX_FILE
+            )
+            # Numba thinks cubinlinker is only needed if the driver is older
+            # than the CUDA runtime, but when PTX files are present, it might
+            # also need to patch because those PTX files may be compiled by
+            # a CUDA version that is newer than the driver as well
+            if (driver_version < ptx_toolkit_version) or (
+                driver_version < runtime_version
+            ):
+                config.CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY = 1
+
+
+def _get_cuda_version_from_ptx_file(path):
+    """
+    https://docs.nvidia.com/cuda/parallel-thread-execution/
+    Each PTX module must begin with a .version
+    directive specifying the PTX language version
+
+    example header:
+    //
+    // Generated by NVIDIA NVVM Compiler
+    //
+    // Compiler Build ID: CL-31057947
+    // Cuda compilation tools, release 11.6, V11.6.124
+    // Based on NVVM 7.0.1
+    //
+
+    .version 7.6
+    .target sm_52
+    .address_size 64
+
+    """
+    with open(path) as ptx_file:
+        for line in ptx_file:
+            if line.startswith(".version"):
+                ver_line = line
+                break
+        else:
+            raise ValueError("Could not read CUDA version from ptx file.")
+    version = ver_line.strip("\n").split(" ")[1]
+    # from ptx_docs/release_notes above:
+    ver_map = {
```

Comment:
How is the list of versions decided on? We support older runtimes and drivers than 11.5. We need to document how this list is made and when it should be modified.

Reply:
These are the versions of NVVM used to produce the PTX files that we ship. At the point at which we introduced the first feature that relied on shipping a PTX file, we were building all of our packages with cuda 11.5. so it was assumed we wouldn't need ever need to look up a version earlier then that. Although I suppose one could imagine building from source on an earlier cuda version. 

I am adding CUDA versions back through 11.0 and 12.1 to be safe. 

We would want to update this every time we start building user facing packages with a new CUDA version that isn't in the table. I am happy to put that info somewhere, not sure where would be best though.

Reply:
Exactly what I was looking for. Summarize that and put it in a comment. 😄

“CUDA 11.5 or higher is required to build cudf. This dictionary maps from supported versions of NVVM to the PTX version it produces.”

I would leave off CUDA versions prior to 11.5, because we cannot use them for other reasons (no int128 support, required for 128-bit decimal types).